core/src/main/scala/reconciler/Reconciler.scala

```diff
@@ -119,6 +119,118 @@ abstract class Reconciler[S <: CustomResource[_, _], T <: ObjectResource] {
+    }
+  }
+
+  def buildResourcesState(implicit context: RequestContext,
+                          sourceFormat: Format[S],
+                          sourceListFormat: Format[ListResource[S]],
+                          sourceResourceDefinition: ResourceDefinition[S],
+                          targetFormat: Format[T],
+                          targetListFormat: Format[ListResource[T]],
+                          targetResourceDefinition: ResourceDefinition[T],
+                          ec: ExecutionContext): Future[ResourcesState] = {
+
+    for {
+      source <- buildResource[S]
+      target <- buildResource[T]
+    } yield
+      ResourcesState(
+        source.watch,
+        target.watch,
+        buildState(source.list, target.list)
+      )
+
+  }
+
+  def scanner(acc: Cache, x: WatchEvent[_ >: T with S <: ObjectResource]): Cache = {
+    val `WatchEvent[S]` = TypeCase[WatchEvent[S]]
+    val `WatchEvent[T]` = TypeCase[WatchEvent[T]]
+
+    x match {
+      case `WatchEvent[S]`(we) =>
+        val key = CacheEntryKey(name = we._object.name, namespace = we._object.namespace, kind = we._object.kind)
+        val targets = acc.cache.get(key).fold(List.empty[T])(_.ts)
+        val cache = acc.cache + (key -> CacheEntry(we._object, targets))
+
+        we._type match {
+          case EventType.ADDED | EventType.MODIFIED => Cache(List(Modified(we._object, targets)), cache)
+          case EventType.DELETED                    => Cache(List(Deleted(we._object, targets)), cache - key)
+        }
+
+      case `WatchEvent[T]`(we) =>
+        val owner = we._object.metadata.ownerReferences.head
+        val sourceKey = CacheEntryKey(name = owner.name, namespace = we._object.namespace, kind = owner.kind)
+
+        acc.cache.get(sourceKey) match {
+          case Some(s) =>
+            val cacheEntry = we._type match {
+              case EventType.DELETED => CacheEntry(s.s, s.ts.filterNot(_.uid == we._object.uid))
+              case _                 => CacheEntry(s.s, List(we._object))
+            }
+            val cache = acc.cache + (sourceKey -> cacheEntry)
+            Cache(List(Modified(s.s, cacheEntry.ts)), cache)
+
+          case None => acc
+        }
+    }
+  }
+
+  def merge(implicit context: RequestContext,
+            sourceFormat: Format[S],
+            sourceListFormat: Format[ListResource[S]],
+            sourceResourceDefinition: ResourceDefinition[S],
+            targetFormat: Format[T],
+            targetListFormat: Format[ListResource[T]],
+            targetResourceDefinition: ResourceDefinition[T],
+            ec: ExecutionContext): akka.stream.scaladsl.Source[List[Event], Future[NotUsed]] = {
+
+    Source.fromFutureSource(
+      for {
+        resourceState <- buildResourcesState
+      } yield
+        Source
+          .combine(resourceState.source, resourceState.target)(Merge(_))
+          .scan(Cache(List.empty, resourceState.state))(scanner)
```

Comment:
Instead of List.empty... why not 
```scala
resourceState.state.values.map(ce => Modified(ce.s, ce.ts))
```
In this way you fix the initial startup problem (i.e this is what `Init` case does within our operator). In other words when the operator starts up it can reconcile the current set of resources to ensure that they are what they should be.

Reply:
Thanks, done.